packages/lib/gpcircuits/src/tuple.ts

```diff
@@ -0,0 +1,81 @@
+import { CircuitSignal } from "./types";
+import { poseidon1, poseidon2, poseidon3, poseidon4} from "poseidon-lite";
+
+const poseidon = [poseidon2, poseidon3, poseidon4];
+
+export type TupleModuleInputs = {
+  valueHash: CircuitSignal[];
+  tupleIndices: CircuitSignal[][];
+};
+
+export type TupleModuleInputNamesType = [
+  "valueHash",
+  "tupleIndices"
+];
+
+export type TupleModuleOutputs = { tupleHashes: CircuitSignal[] };
+
+export type TupleModuleOutputNamesType = ["tupleHashes"];
+
+export class Tuple {
+  /**
+   * Computes the hash representation of tuples formed from value hashes
+   * and previously computed value hashes (cf. `TupleModule`). Matches
+   * the implementation of `TupleModule` and assumes appropriate padding
+   * (i.e. `BABY_JUB_NEGATIVE_ONE` for unused tuple indices) and array
+   * lengths (i.e. all tuple index tuples must have length matching the
+   * tuple arity parameter).
+   *
+   * @param `paramTupleArity` denotes the tuple arity.
+   * @param `paramMaxValues` denotes the maximum number of POD entry values.
+   * @param `paramMaxTuples` denotes the maximum number of tuples to form.
+   * @param `valueHash` denotes the array of POD entry value hashes.
+   * @param `tupleIndices` denotes the list of tuples of indices of
+   *  arity `paramTupleArity`, where an index `i` less than `paramMaxValues`
+   *  refers to a POD object value with that index and `i >= paramMaxValues`
+   *  refers to the `i - paramMaxValues`th tuple.
+   * @returns `bigint[]` of tuple hashes.
+   * @throws TypeError if the given tuple arity does not lie within [2,4].
```

Comment:
I don't think it's necessary to call out all the individual @throws scenarios here.  I don't mind, but it might be hard to keep up-to-date when code changes.  I think it's sufficient to say `@throws TypeError if inputs are out of range`.

Actually shouldn't that be a RangeError?